internal/wrphandlers/qos/priority_queue.go

```diff
@@ -0,0 +1,103 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"container/heap"
+
+	"github.com/xmidt-org/wrp-go/v3"
+)
+
+// PriorityQueue implements heap.Interface and holds wrp Message, using wrp.QOSValue as its priority.
+// https://xmidt.io/docs/wrp/basics/#qos-description-qos
+type PriorityQueue struct {
+	// heap for wrp messages, ingested by serviceQOS
+	heap []wrp.Message
+	// maxHeapSize is the allowable max size of the queue based on the sum of all queued wrp message's payloads
+	maxHeapSize int
+}
+
+// Dequeue returns the next highest priority message.
+func (pq *PriorityQueue) Dequeue() (wrp.Message, bool) {
+	// Required, otherwise heap.Pop will panic during an internal Swap call.
+	if pq.Len() == 0 {
+		return wrp.Message{}, false
+	}
+
+	msg, ok := heap.Pop(pq).(wrp.Message)
+
+	return msg, ok
+}
+
+// Enqueue queues the given message.
+func (pq *PriorityQueue) Enqueue(msg wrp.Message) {
+	// Check whether msg would already violate maxHeapSize.
+	if len(msg.Payload) >= pq.maxHeapSize {
```

Comment:
```suggestion
	if len(msg.Payload) > pq.maxHeapSize {
```

Reply:
@johnabass besides checking  whether msg would already violate `maxHeapSize`, should we consider enforcing a max size for a given message?

Otherwise, a non-empty queue may be emptied out in order to queue an incoming message (with a similar or great `QOS` value) of size `maxHeapSize`

Reply:
I would have a max message size as well.  It's generally a best practice to bound both (1) the number of units of work, and (2) the size of a single unit of work.

That does leave you with a design decision though:  What do you do with messages that are too large?  How do you indicate that to the caller?  There's lots of right ways to do that.